File: xls/fuzzer/sample_generator_test.py
# Lint as: python3
#
# Copyright 2020 The XLS Authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import collections
import random

from xls.common import test_base
from xls.dslx.python.cpp_concrete_type import ArrayType
from xls.dslx.python.cpp_concrete_type import BitsType
from xls.dslx.python.cpp_concrete_type import TupleType
from xls.fuzzer import ast_generator
from xls.fuzzer import sample_generator
from xls.fuzzer.sample import SampleOptions


class SampleGeneratorTest(test_base.TestCase):

  def test_randrange_biased_towards_zero(self):
    rng = random.Random(0)
    counter = collections.Counter()
    for _ in range(1024):
      counter[sample_generator.randrange_biased_towards_zero(16, rng)] += 1
    print(sorted(counter.items()))
    self.assertTrue(all(k < 16 for k in counter.keys()))
    self.assertTrue(all(counter[0] > counter[i] for i in range(1, 16)))

  def test_generate_empty_arguments(self):
    rng = random.Random(0)
    self.assertEqual(sample_generator.generate_arguments((), rng), ())

  def test_generate_single_bits_arguments(self):
    rng = random.Random(0)
    args = sample_generator.generate_arguments(
        (BitsType(signed=False, size=42),), rng)
    self.assertLen(args, 1)
    self.assertTrue(args[0].is_ubits())
    self.assertEqual(args[0].get_bit_count(), 42)

  def test_generate_mixed_bits_arguments(self):
    rng = random.Random(0)
    args = sample_generator.generate_arguments(
        (BitsType(signed=False, size=123), BitsType(signed=True, size=22)), rng)
    self.assertLen(args, 2)
    self.assertTrue(args[0].is_ubits())
    self.assertEqual(args[0].get_bit_count(), 123)
    self.assertTrue(args[1].is_sbits())
    self.assertEqual(args[1].get_bit_count(), 22)

  def test_generate_tuple_argument(self):
    rng = random.Random(0)
    args = sample_generator.generate_arguments((TupleType(
        (BitsType(signed=False, size=123),
         BitsType(signed=True, size=22))),), rng)
    self.assertLen(args, 1)
    self.assertTrue(args[0].is_tuple())
    self.assertEqual(args[0].get_elements()[0].get_bit_count(), 123)
    self.assertEqual(args[0].get_elements()[1].get_bit_count(), 22)

  def test_generate_array_argument(self):
    rng = random.Random(0)
    args = sample_generator.generate_arguments(
        (ArrayType(BitsType(signed=True, size=4), 24),), rng)
    self.assertLen(args, 1)
    self.assertTrue(args[0].is_array())
    self.assertLen(args[0].get_elements(), 24)
    self.assertTrue(args[0].index(0).is_sbits())
    self.assertTrue(args[0].index(0).get_bit_count(), 4)

  def test_generate_basic_sample(self):
    rng = random.Random(0)
    sample = sample_generator.generate_sample(
        rng,
        ast_generator.AstGeneratorOptions(),
        calls_per_sample=3,
        default_options=SampleOptions(
            convert_to_ir=True, optimize_ir=True, codegen=False,
            simulate=False))
    self.assertTrue(sample.options.input_is_dslx)
    self.assertTrue(sample.options.convert_to_ir)
    self.assertTrue(sample.options.optimize_ir)
    self.assertFalse(sample.options.codegen)
    self.assertFalse(sample.options.simulate)
    self.assertLen(sample.args_batch, 3)
    self.assertIn('fn main', sample.input_text)

  def test_generate_codegen_sample(self):
    rng = random.Random(0)
    sample = sample_generator.generate_sample(
        rng,
        ast_generator.AstGeneratorOptions(),
        calls_per_sample=0,
        default_options=SampleOptions(
            convert_to_ir=True, optimize_ir=True, codegen=True, simulate=True))
    self.assertTrue(sample.options.input_is_dslx)
    self.assertTrue(sample.options.convert_to_ir)
    self.assertTrue(sample.options.optimize_ir)
    self.assertTrue(sample.options.codegen)
    self.assertTrue(sample.options.simulate)
    self.assertNotEmpty(sample.options.codegen_args)
    self.assertEmpty(sample.args_batch)


if __name__ == '__main__':
  test_base.main()

File: xls/fuzzer/sample_generator_test.cc
// Copyright 2021 The XLS Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "xls/fuzzer/sample_generator.h"

#include <cstdint>
#include <memory>
#include <random>
#include <string>
#include <vector>

#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "absl/status/status_matchers.h"
#include "xls/common/status/matchers.h"
#include "xls/dslx/channel_direction.h"
#include "xls/dslx/frontend/pos.h"
#include "xls/dslx/interp_value.h"
#include "xls/dslx/type_system/type.h"
#include "xls/fuzzer/ast_generator.h"
#include "xls/fuzzer/sample.h"
#include "xls/fuzzer/value_generator.h"

namespace xls {
namespace {

using ::absl_testing::IsOkAndHolds;
using ::testing::HasSubstr;

TEST(SampleGeneratorTest, GenerateBasicFunctionSample) {
  dslx::FileTable file_table;
  std::mt19937_64 rng;
  SampleOptions sample_options;
  constexpr int kCallsPerSample = 3;
  sample_options.set_calls_per_sample(kCallsPerSample);
  XLS_ASSERT_OK_AND_ASSIGN(
      Sample sample, GenerateSample(dslx::AstGeneratorOptions{}, sample_options,
                                    rng, file_table));
  EXPECT_TRUE(sample.options().input_is_dslx());
  EXPECT_TRUE(sample.options().convert_to_ir());
  EXPECT_TRUE(sample.options().optimize_ir());
  EXPECT_FALSE(sample.options().codegen());
  EXPECT_FALSE(sample.options().simulate());

  std::vector<std::vector<dslx::InterpValue>> args_batch;
  XLS_EXPECT_OK(sample.GetArgsAndChannels(args_batch));
  EXPECT_EQ(args_batch.size(), kCallsPerSample);
  EXPECT_THAT(sample.input_text(), testing::HasSubstr("fn main"));
}

TEST(SampleGeneratorTest, GenerateCodegenSample) {
  dslx::FileTable file_table;
  std::mt19937_64 rng;
  SampleOptions sample_options;
  sample_options.set_codegen(true);
  sample_options.set_simulate(true);
  constexpr int64_t kCallsPerSample = 0;
  sample_options.set_calls_per_sample(kCallsPerSample);
  XLS_ASSERT_OK_AND_ASSIGN(
      Sample sample, GenerateSample(dslx::AstGeneratorOptions{}, sample_options,
                                    rng, file_table));
  EXPECT_TRUE(sample.options().input_is_dslx());
  EXPECT_TRUE(sample.options().convert_to_ir());
  EXPECT_TRUE(sample.options().optimize_ir());
  EXPECT_TRUE(sample.options().codegen());
  EXPECT_TRUE(sample.options().simulate());
  EXPECT_FALSE(sample.options().codegen_args().empty());
  std::vector<std::vector<dslx::InterpValue>> args_batch;
  XLS_EXPECT_OK(sample.GetArgsAndChannels(args_batch));
  EXPECT_EQ(args_batch.size(), kCallsPerSample);
}

TEST(SampleGeneratorTest, GenerateChannelArgument) {
  std::mt19937_64 rng;
  std::vector<std::unique_ptr<dslx::Type>> param_types;
  constexpr int64_t kBitCount = 4;
  param_types.push_back(
      std::make_unique<dslx::ChannelType>(std::make_unique<dslx::BitsType>(
                                              /*signed=*/true,
                                              /*size=*/kBitCount),
                                          dslx::ChannelDirection::kOut));

  std::vector<const dslx::Type*> param_type_ptrs;
  param_type_ptrs.reserve(param_types.size());
  for (const auto& t : param_types) {
    param_type_ptrs.push_back(t.get());
  }
  XLS_ASSERT_OK_AND_ASSIGN(std::vector<dslx::InterpValue> arguments,
                           GenerateInterpValues(rng, param_type_ptrs));
  ASSERT_EQ(arguments.size(), 1);
  ASSERT_EQ(arguments.size(), param_types.size());
  const dslx::InterpValue& value = arguments[0];
  ASSERT_TRUE(value.IsSBits());
  EXPECT_THAT(value.GetBitCount(), IsOkAndHolds(kBitCount));
}

TEST(SampleGeneratorTest, GenerateBasicProcSample) {
  dslx::FileTable file_table;
  std::mt19937_64 rng;
  SampleOptions sample_options;
  constexpr int64_t kProcTicks = 3;
  sample_options.set_sample_type(fuzzer::SampleType::SAMPLE_TYPE_PROC);
  sample_options.set_calls_per_sample(0);
  sample_options.set_proc_ticks(kProcTicks);
  XLS_ASSERT_OK_AND_ASSIGN(
      Sample sample,
      GenerateSample(dslx::AstGeneratorOptions{.generate_proc = true},
                     sample_options, rng, file_table));
  EXPECT_TRUE(sample.options().input_is_dslx());
  EXPECT_TRUE(sample.options().convert_to_ir());
  EXPECT_TRUE(sample.options().optimize_ir());
  EXPECT_FALSE(sample.options().codegen());
  EXPECT_FALSE(sample.options().simulate());

  std::vector<std::vector<dslx::InterpValue>> args_batch;
  std::vector<std::string> ir_channel_names;
  XLS_EXPECT_OK(sample.GetArgsAndChannels(args_batch, &ir_channel_names));
  EXPECT_EQ(args_batch.size(), kProcTicks);

  EXPECT_THAT(sample.input_text(), HasSubstr("proc main"));
}

}  // namespace
}  // namespace xls
